lib/spack/spack/solver/asp.py

```diff
@@ -1873,6 +1874,12 @@ def define_version_constraints(self):
             # generate facts for each package constraint and the version
             # that satisfies it
             for v in allowed_versions:
+                # TODO: with the below 3 lines, the latter check in
+                # test_requirement_adds_multiple_new_versions passes, but
+                # this doesn't seem like the right thing to do...
+                # if isinstance(v, spack.version.GitVersion) and v.user_supplied_reference:
+                #     x, _ = str(v).split("=")
+                #     self.gen.fact(fn.version_satisfies(pkg_name, x, v))
```

Comment:
Right now the ASP facts look like:

```
condition_requirement(0,"node_version_satisfies","v","commits[1]=2.2").
imposed_constraint(0,"node_version_satisfies","v","commits[1]=2.2").

%============================================================================
% Version Constraints
%============================================================================
version_satisfies("v","commits[1]","commits[1]").
version_satisfies("v","commits[1]=2.2","commits[1]=2.2").
version_satisfies("v","commits[1]=2.3","commits[1]=2.3").
```

Which looks strange to me. I wrote this in an attempt to generate:

```
version_satisfies("v","commits[1]","commits[1]=2.3").
```

It feels like these `hash=numbered-version` strings are being treated like atomic pieces of data though, which seems odd to me.

Reply:
Partial update: I've broken down the versions as they are entered into the solver, such that you don't have atoms like "hash=number". The resulting challenge is restoring that information to the specs after building them from the solution. For now I've resolved that when `hash=number` is given on the command line, but if it exists in the `requires:` section, that would involve extra effort.